src/umessage/umessagebuilder.rs

```diff
@@ -276,31 +278,87 @@ impl<'a> UMessageBuilder<'a> {
     /// # Ok(())
     /// # }
     /// ```
-    pub fn response_for_request(request_attributes: &'a UAttributes) -> UMessageBuilder<'a> {
+    pub fn response_for_request(request_attributes: &UAttributes) -> UMessageBuilder {
         UMessageBuilder {
             validator: Box::new(ResponseValidator),
             message_type: UMessageType::UMESSAGE_TYPE_RESPONSE,
-            source: request_attributes.sink.as_ref(),
-            sink: request_attributes.source.as_ref(),
-            request_id: request_attributes.reqid.as_ref(),
+            source: request_attributes.sink.as_ref().cloned(),
+            sink: request_attributes.source.as_ref().cloned(),
+            request_id: request_attributes.id.as_ref().cloned(),
             ..Default::default()
         }
     }
 
+    /// Sets the message's identifier.
+    ///
+    /// Every message must have an identifier and any of the build functions will return an error if no
+    /// identifier has been set.
+    ///
+    /// This function is particularly helpful in order to re-use a builder for creating multiple
+    /// messages consecutively based on the same attributes, differing in payload only as illustrated
+    /// in the examples.
+    ///
+    /// # Arguments
+    ///
+    /// * `message_id` - The identifier to use.
+    ///
+    /// # Returns
+    ///
+    /// The builder.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given UUID is not a [valid uProtocol UUID](`UUID::is_uprotocol_uuid`).
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use up_rust::{UMessageBuilder, UMessageType, UPayloadFormat, UPriority, UUIDBuilder, UUri};
+    ///
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// let uuid_builder = UUIDBuilder::new();
+    /// let topic = UUri::try_from("my-vehicle/cabin/1/doors.driver_side#status")?;
+    /// let mut builder = UMessageBuilder::publish(topic);
+    /// builder.with_priority(UPriority::UPRIORITY_CS2);
+    /// let message_one = builder
+    ///                     .with_message_id(uuid_builder.build())
+    ///                     .build_with_payload("closed".into(), UPayloadFormat::UPAYLOAD_FORMAT_TEXT)?;
+    /// let message_two = builder
+    ///                     // use new message ID but retain all other attributes
+    ///                     .with_message_id(uuid_builder.build())
+    ///                     .build_with_payload("open".into(), UPayloadFormat::UPAYLOAD_FORMAT_TEXT)?;
+    /// assert_ne!(message_one.attributes.id, message_two.attributes.id);
+    /// assert_eq!(message_one.attributes.source, message_two.attributes.source);
+    /// assert_eq!(message_one.attributes.priority, UPriority::UPRIORITY_CS2.into());
+    /// assert_eq!(message_two.attributes.priority, UPriority::UPRIORITY_CS2.into());
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn with_message_id(&mut self, message_id: UUID) -> &mut UMessageBuilder {
```

Comment:
I really like your proposal to be able to reuse `UMessageBuilder` in client code.

One pattern that I've seen used elsewhere is to have an 'intermediate' builder and then when the builder has been 'finalized' only then allow to call `build()`. Can help prevent building a message that doesn't have what needed.

So, for example, up till we call `with_message_id()`, the type could be `UMessageBuilder` and then after both of those are called and we have `message_id` set, we could return a type `UMessageBuildable` (that'd be temporary, ephemeral).

And then we'd impl `build_with_payload()` only on `UMessageBuildable`.

What do you think? Kind of only gave a rough sketch